### PR TITLE
Relocate pointer in temporary segments

### PR DESCRIPTION
**Pull request by pefontana**
# Relocate pointer in temporary segments

Add the feature to relocate temporary segments

## Checklist
- [ ] Linked to Github Issue
- [x] Unit tests added
- [ ] Integration tests added.
- [ ] This change requires new documentation.
  - [ ] Documentation has been added/updated.

### CODE REVIEW

File: src/types/relocatable.rs [hunk 1]
```diff
@@ -354,14 +354,21 @@ impl<'a> Add<usize> for &'a Relocatable {
 pub fn relocate_value(
     value: MaybeRelocatable,
     relocation_table: &Vec<usize>,
+    relocation_rules: &HashMap<usize, Relocatable>,
```
Comment (Oppen) [inline src/types/relocatable.rs:357]: It doesn't look like we're ever using the `.offset` field, so maybe just store a `usize` for the index here?

File: src/types/relocatable.rs [hunk 2]
```diff
@@ -354,14 +354,21 @@
...
     match value {
         MaybeRelocatable::Int(num) => Ok(num),
         MaybeRelocatable::RelocatableValue(relocatable) => {
-            let segment_index: usize = relocatable
-                .segment_index
-                .try_into()
-                .map_err(|_| MemoryError::AddressInTemporarySegment(relocatable.segment_index))?;
+            let segment_index = if relocatable.segment_index.is_negative() {
+                relocation_rules
+                    .get(&(relocatable.segment_index.abs_diff(0)))
```
Comment (Oppen) [inline src/types/relocatable.rs:364]: I think this is the same as `.segment_index.abs()`.

**Review by Oppen — COMMENTED** [review]
LGTM, some minor stuff tho.

File: src/types/relocatable.rs (hunk 1 above)
Comment (pefontana) [reply src/types/relocatable.rs:357]: Thanks for this one @Oppen. 
I found this a little bit strange, so I made some testing and found that we are having a bug in some edge cases:
If we relocate a temporary segment to an address with offset != 0, we are having some differences with the Python VM `relocate_segment(src_ptr=temporary_array, dest_ptr=array + 5)`
I think we need to make some modifications to the Memory::relocate_value and the Relocatable::relocate_value
I will tackle this issue as soon as we merge the #576

Comment (pefontana) [reply src/types/relocatable.rs:357]: https://github.com/lambdaclass/cairo-rs/issues/581

**Review by Oppen — APPROVED** [review]
LGTM

**Review by fmoletta — APPROVED** [review]

**Pull request merged**